mpc-spec/src/circuit.rs

```diff
@@ -0,0 +1,307 @@
+/// evaluation.
+pub type InputWidth = usize;
+/// Representation of a circuit evaluated by an MPC engine.
+#[derive(Debug, Clone)]
+pub struct Circuit {
+    /// The bit-width of the inputs expected by the different parties,
+    /// [InputWidth] at index `i` representing the number of input bits for
+    /// party `i`.
+    pub input_widths: Vec<InputWidth>,
+    /// The circuit's gates.
+    pub gates: Vec<WiredGate>,
+    /// The indices of the gates in [`Circuit::gates`] that produce output bits.
+    pub output_gates: Vec<WireIndex>,
+}
+
+/// Errors occurring during the validation or the execution of the MPC protocol.
+#[derive(Debug, PartialEq, Eq)]
+pub enum CircuitError {
+    /// The provided party input does not match the number of input bits for
+    /// that party expected by the circuit.
+    PartyInputMismatch(usize, usize),
+    /// The provided set of inputs does not match the number of party inputs
+    /// expected by the circuit.
+    PartyCountMismatch(usize, usize),
+    /// The gate with the specified wire index contains invalid gate connections
+    /// or is placed out of sequence.
+    InvalidGate(usize),
+    /// The specified output gate does not exist in the circuit.
+    InvalidOutputWire(usize),
+    /// The circuit does not specify any output gates.
+    EmptyOutputSpecification,
+    /// The circuit does not specify input wires.
+    EmptyInputSpecification,
+    /// The circuit specifies a zero-width input.
+    InvalidInputSpecification,
+}
+
+impl std::fmt::Display for CircuitError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CircuitError::PartyInputMismatch(expected_inputs, actual_inputs) => write!(
+                f,
+                "Expected {} input bits for a party, but received {} input bits.",
+                *expected_inputs, *actual_inputs
+            ),
+            CircuitError::PartyCountMismatch(expected_parties, actual_parties) => write!(
+                f,
+                "Expected inputs for {} parties, but received inputs for {} parties.",
+                *expected_parties, *actual_parties
+            ),
+            CircuitError::InvalidGate(gate_index) => write!(
+                f,
+                "Gate {}: Out of order placement or invalid wiring.",
+                *gate_index
+            ),
+            CircuitError::InvalidOutputWire(oob_index) => {
+                write!(f, "Output index out of bounds: {}", *oob_index)
+            }
+            CircuitError::EmptyOutputSpecification => {
+                write!(f, "Circuit does not specify output bits.")
+            }
+            CircuitError::EmptyInputSpecification => {
+                write!(f, "Circuit does not specify any party inputs.")
+            }
+            CircuitError::InvalidInputSpecification => {
+                write!(f, "Circuit specifies an empty party input.")
+            }
+        }
+    }
+}
```

Comment:
I am not sure if we have a clear policy here, but the [documentation of the standard library Error trait](https://doc.rust-lang.org/stable/std/error/trait.Error.html) says

> Errors must describe themselves through the [Display](https://doc.rust-lang.org/stable/std/fmt/trait.Display.html) and [Debug](https://doc.rust-lang.org/stable/std/fmt/trait.Debug.html) traits. Error messages are typically concise lowercase sentences without trailing punctuation

Personally I think it makes sense to follow that convention, but like I said, not sure if we have a policy for this.

Reply:
AFAIK we don't have our own policy here, so it makes sense to follow the standard library approach. Thanks for pointing this out!